rep-2004.rst

```diff
@@ -0,0 +1,836 @@
+However, a package which provides pose estimation (like ``robot_pose_ekf``\ ) is a generic solution for something that most people need, but is often replaced with a domain specific solution in production, and therefore it should probably not start out as 'Level 1'.
+Such a package may upgrade to 'Level 1' at a later date, if it proves to be a solution that people want to use in their products.
+
+Tools, like ``rostopic``\ , generally do not fall into this category either, but are not categorically excluded.
+For example, it may be the case that the tool which launches and verifies a ROS graph (``ros2launch``\ ) may need to be considered 'Level 1' for use in production systems.
+
+Package Requirements
+~~~~~~~~~~~~~~~~~~~~
+
+Requirements to be considered a 'Level 1' package:
+
+.. _Version Policy:
+
+1. **Version Policy:**
+
+   .. _1.i:
+
+   i. Must have a version policy (e.g. ``semver``)
+
+   .. _1.ii:
+
+   ii. Must be at a stable version (e.g. for ``semver`` that means version >= 1.0.0)
+
+   .. _1.iii:
+
+   iii. Must have a strictly declared public API
+
+   .. _1.iv:
+
+   iv. Must have a policy for API stability
+
+   .. _1.v:
+
+   v. Must have a policy for ABI stability
+
+   .. _1.vi:
+
+   vi. Must have a policy that keeps API and ABI stability within a released ROS distribution
+
+.. _Change Control Process:
+
+2. **Change Control Process:**
+
+   .. _2.i:
+
+   i. Must have all code changes occur through a change request (e.g. pull request, merge request, etc.)
+
+   .. _2.ii:
+
+   ii. Must have confirmation of contributor origin (e.g. `DCO  <https://developercertificate.org/>`_, CLA, etc.)
+
+   .. _2.iii:
+
+   iii. Must have peer review policy for all change requests (e.g. require one or more reviewers)
+
+   .. _2.iv:
+
+   iv. Must have Continuous Integration (CI) policy for all change requests
+
+   .. _2.v:
+
+   v. Must have documentation policy for all change requests
+
+.. _Documentation:
+
+3. **Documentation:**
+
+   .. _3.i:
+
+   i. Must have documentation for each "feature" (e.g. for ``rclcpp``: create a node, publish a message, spin, etc.)
+
+   .. _3.ii:
+
+   ii. Must have documentation for each item in the public API (e.g. functions, classes, etc.)
+
+   .. _3.iii:
+
+   iii. Must have a declared license or set of licenses
+
+   .. _3.iv:
+
+   iv. Must have a copyright statement in each source file
+
+   .. _3.v:
+
+   v. Must have a "quality declaration" document, which declares the quality level and justifies how the package meets each of the requirements
+
+      .. _3.v.a:
+
+      a. *Must have a section in the repository's ``README`` which contains the "quality declaration" or links to it*
+
+      .. _3.v.b:
+
+      b. *Must register with a centralized list of 'Level N' packages, if one exists, to allow for peer review of the claim*
+
+.. _Testing:
+
+4. **Testing:**
```

Comment:
What is the testing policy for things like threadsafety?

Reply:
I don't think we need anything specific to that, to be honest. The documentation should probably state what the thread-safety status of the methods are, and the tests should aim to check that documented behavior.

I think something like this is out of the scope of this REP because I think (i.e. my opinion) it strays into "write good software" which always tends to be subjective and hard to completely quantify in my experience. Like how do you measure when enough thread-safety testing has been done, does it need to proven somehow (logical proof), is it empirical testing? I don't know. Also, why just thread-safety, why not memory allocations, or blocking behavior, or some other aspect of the software.

If you have some specific in mind, please suggest what you think a reasonable entry here would be and maybe we can agree on something generic and useful enough to include here.

It may also be something good to have in our developer guide or in a contributing or pull request review check list, depending on what you were thinking.

Reply:
> I think something like this is out of the scope of this REP because I think (i.e. my opinion) it strays into "write good software" which always tends to be subjective and hard to completely quantify in my experience.

I agree that this is REALLY hard to nail down in a prescriptive way, and quite possibly belongs elsewhere. It also likely doesn't belong under **Testing** unless and until the community has had more experience ensuring threadsafe code.

> If you have some specific in mind, please suggest what you think a reasonable entry here would be and maybe we can agree on something generic and useful enough to include here.

Challenge accepted!

**Platform Support**:
   ii. Should document additional constraints on the runtime environment, including whether APIs may be used multithreaded, whether filesystem access is required, and what additional sensors or hardware are expected.

Reply:
At the risk of loosing momentum on this point, I want to wrap up this first version of the REP. I will make a follow up issue for this and if you have time a pull request with what you've suggested would be easier to comment on (and make sure things like: it is considered or not for each quality level, and it is in the table of requirements, etc...).